packages/utils/src/abi_utils.ts

```diff
@@ -1,7 +1,166 @@
+        } else {
+            names.push(param.name || null);
+            types.push(param.type);
+        }
+    });
+
+    return {
+        names,
+        types,
+    };
+}
+
+// returns true if x is equal to y and false otherwise. Performs some minimal
+// type conversion and data massaging for x and y, depending on type. name and
+// type should typically be derived from parseEthersParams.
+function isAbiDataEqual(name: EthersParamName, type: string, x: any, y: any): boolean {
+    if (_.isUndefined(x) && _.isUndefined(y)) {
+        return true;
+    } else if (_.isUndefined(x) && !_.isUndefined(y)) {
+        return false;
+    } else if (!_.isUndefined(x) && _.isUndefined(y)) {
+        return false;
+    }
+    if (_.endsWith(type, '[]')) {
+        // For array types, we iterate through the elements and check each one
+        // individually. Strangely, name does not need to be changed in this
+        // case.
+        if (x.length !== y.length) {
+            return false;
+        }
+        const newType = _.trimEnd(type, '[]');
+        for (let i = 0; i < x.length; i++) {
+            if (!isAbiDataEqual(name, newType, x[i], y[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+    if (_.startsWith(type, 'tuple(')) {
+        if (_.isString(name)) {
+            throw new Error('Internal error: type was tuple but names was a string');
+        } else if (_.isNull(name)) {
+            throw new Error('Internal error: type was tuple but names was a null');
+        }
+        // For tuples, we iterate through the underlying values and check each
+        // one individually.
+        const types = splitTupleTypes(type);
+        if (types.length !== name.names.length) {
+            throw new Error(
+                `Internal error: parameter types/names length mismatch (${types.length} != ${name.names.length})`,
+            );
+        }
+        for (let i = 0; i < types.length; i++) {
+            // For tuples, name is an object with a names property that is an
+            // array. As an example, for orders, name looks like:
+            //
+            //  {
+            //      name: 'orders',
+            //      names: [
+            //          'makerAddress',
+            //          // ...
+            //          'takerAssetData'
+            //      ]
+            //  }
+            //
+            const nestedName = _.isString(name.names[i])
+                ? (name.names[i] as string)
+                : ((name.names[i] as EthersNestedParamName).name as string);
+            if (!isAbiDataEqual(name.names[i], types[i], x[nestedName], y[nestedName])) {
+                return false;
+            }
+        }
+        return true;
+    } else if (type === 'address' || type === 'bytes') {
+        // HACK(albrow): ethers.js sometimes changes the case of addresses/bytes
+        // when decoding/encoding. To account for that, we convert to lowercase
```

Comment:
Do you mean that ethers.js is modifying by reference? If so, could you submit an issue and add the source link here? That way, once it's fixed, we could remove this logic.

Reply:
I see. Yep, I'd just mention that Ethers always returns checksummed addresses, even when initially supplied a non-checksummed address.